### PR TITLE
Deal Markers on Map

### PR DESCRIPTION
**Pull request by kgunadhi**
Map has a marker for each deal that shows the store name and deal title, and tapping on the info window segues to the details view for that deal.
Screen Shot 2020-07-30 at 9 26 46 AM

### CODE REVIEW

File: DiscountDiscover/Models/Deal.h [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 #import <Foundation/Foundation.h>
 #import <CoreLocation/CoreLocation.h>
+#import <GoogleMaps/GoogleMaps.h>
```
Comment (irodrigo17) [inline DiscountDiscover/Models/Deal.h:11]: This model should not be coupled with an external SDK. Instead you could create a `DealMarker` class that has a reference to the deal and another one to the marker.

File: DiscountDiscover/Models/Deal.m [hunk 2]
```diff
@@ -69,6 +71,13 @@ - (NSString *)formatStoreAddressString:(NSDictionary *)dictionary {
     }
 }
 
+- (GMSMarker *)createMarker {
+    GMSMarker *marker = [GMSMarker markerWithPosition:self.storeCoordinate];
```
Comment (irodrigo17) [inline DiscountDiscover/Models/Deal.m:79]: As a stretch goal, it could be nice to easily differentiate deals on the map by using different marker colors, for example depending on the discount % or on the deal category.

File: DiscountDiscover/View Controllers/MapViewController.h [hunk 3]
```diff
@@ -7,11 +7,15 @@
 //
 
 #import <UIKit/UIKit.h>
+#import <GoogleMaps/GoogleMaps.h>
```
Comment (irodrigo17) [inline DiscountDiscover/View Controllers/MapViewController.h:10]: Why did you move this dependency from the .m to the to the .h file? Ideally we only want to import dependencies in the header file when strictly necessary, but should prefer to use forward declarations when possible (using `@class` and `@protocol` instead of `#import`).

File: DiscountDiscover/View Controllers/MapViewController.m [hunk 4]
```diff
@@ -24,10 +25,18 @@ - (void)viewDidLoad {
...
-    GMSMapView *mapView = [GMSMapView mapWithFrame:self.view.frame camera:camera];
-    mapView.settings.myLocationButton = YES;
-    mapView.myLocationEnabled = YES;
-    [self.view addSubview:mapView];
+    self.mapView = [GMSMapView mapWithFrame:self.view.frame camera:camera];
+    self.mapView.settings.myLocationButton = YES;
+    self.mapView.myLocationEnabled = YES;
+    [self.view addSubview:self.mapView];
+    
+    [self addDealMarkers];
```
Comment (irodrigo17) [inline DiscountDiscover/View Controllers/MapViewController.m:35]: If you only add the markers in `viewDidLoad` they won't get updated according to the list. Seems like you might want to move this to viewWillAppear or something later so the map always shows the same as the list. 
Also, are you planning to fetch new deals when user scrolls or zooms the map? In this case, would you also update the list view accordingly or should they be independent?

File: DiscountDiscover/View Controllers/MapViewController.m [hunk 5]
```diff
@@ -39,4 +42,22 @@ - (void)addDealMarkers {
     }
 }
 
+- (void)mapView:(GMSMapView *)mapView didTapInfoWindowOfMarker:(GMSMarker *)marker {
+    
+    NSUInteger i = [self.deals indexOfObjectPassingTest:^BOOL(Deal *deal, NSUInteger idx, BOOL *stop) {
+        return [deal.name isEqual:marker.snippet];
```
Comment (irodrigo17) [inline DiscountDiscover/View Controllers/MapViewController.m:47]: You shouldn't compare objects by name/description, use an ID instead. You can also compare objects directly by memory address if possible to make sure they're exactly the same object. Adding a `DealMarker` class as I suggested before would also help here by giving you direct access to the deal from the marker. You should also check if you can attach a random object to the marker besides the snippet and title, if you could do something like `marker.object = DealMarkerInstance` then you have direct access to the deal without iterating though an array.

**Review by irodrigo17 — COMMENTED** [review]
Looks good overall, just a few comments. 
Thanks for addressing all previous feedback!

File: DiscountDiscover/View Controllers/MapViewController.m (hunk 4 above)
Comment (kgunadhi) [reply DiscountDiscover/View Controllers/MapViewController.m:35]: Ah I realized and fixed this in a commit right after this PR unfortunately! And I wasn't planning on it, but that would definitely be a good thing to look into next week.

File: DiscountDiscover/View Controllers/MapViewController.m (hunk 5 above)
Comment (irodrigo17) [reply DiscountDiscover/View Controllers/MapViewController.m:47]: See the [`userData`](https://developers.google.com/maps/documentation/ios-sdk/reference/interface_g_m_s_overlay#a900ff2be989386d501ed6c64ea1d7ab3) property, you can set that to the `Deal` instance that the marker represents and avoid the iteration and comparing deals.
